src/ssram.c

```diff
@@ -148,53 +164,96 @@ void ice_ssram_deinit(void) {
         hw_set_bits(&dma_hw->ch[g_rx_dma_channel].al1_ctrl, DMA_CH0_CTRL_TRIG_INCR_WRITE_BITS);
-        dma_channel_transfer_to_buffer_now(g_rx_dma_channel, dest, size);
+        dma_channel_transfer_to_buffer_now(g_rx_dma_channel, data, data_size);
 
         hw_clear_bits(&dma_hw->ch[g_tx_dma_channel].al1_ctrl, DMA_CH0_CTRL_TRIG_INCR_READ_BITS);
-        dma_channel_transfer_from_buffer_now(g_tx_dma_channel, &g_dummy, size);
+        dma_channel_transfer_from_buffer_now(g_tx_dma_channel, &g_dummy, data_size);
     } else {
-        spi_read_blocking(SPI_SSRAM, 0, dest, size);
+        spi_read_blocking(SPI_SSRAM, 0, data, data_size);
         ice_ssram_deselect();
     }
 }
+
+uint8_t ice_ssram_get_status(int cs_pin) {
+    const uint8_t command[] = { CMD_STATUS };
+    uint8_t status;
+    ice_ssram_input_command(cs_pin, command, sizeof(command), &status, sizeof(status), false);
+    return status;
+}
+
+void ice_ssram_erase_chip(int cs_pin) {
+    const uint8_t command[] = { CMD_CHIP_ERASE };
+    ice_ssram_output_command(cs_pin, command, sizeof(command), NULL, 0, false);
+}
+
+void ice_ssram_erase_sector(int cs_pin, uint32_t dest_addr) {
+    assert(addr % ICE_FLASH_PAGE_SIZE == 0);
+
+    const uint8_t command[] = { CMD_SECTOR_ERASE, dest_addr >> 16, dest_addr >> 8, dest_addr };
+    ice_ssram_output_command(cs_pin, command, sizeof(command), NULL, 0, false);
+}
+
+void ice_ssram_enable_write(int cs_pin, bool enabled) {
+    const uint8_t command[] = { enabled ? CMD_ENABLE_WRITE : CMD_DISABLE_WRITE };
+    ice_ssram_input_command(cs_pin, command, sizeof(command), NULL, 0, false);
+}
+
+void ice_ssram_write(int cs_pin, uint32_t dest_addr, const void* src, uint32_t size) {
+    uint8_t command[] = { CMD_WRITE, dest_addr >> 16, dest_addr >> 8, dest_addr };
+    ice_ssram_output_command(cs_pin, command, sizeof(command), src, size, false);
+}
+
+void ice_ssram_write_async(int cs_pin, uint32_t dest_addr, const void* src, uint32_t size) {
+    uint8_t command[] = { CMD_WRITE, dest_addr >> 16, dest_addr >> 8, dest_addr };
+    ice_ssram_output_command(cs_pin, command, sizeof(command), src, size, true);
+}
+
+void ice_ssram_read(int cs_pin, void* dest, uint32_t src_addr, uint32_t size) {
+    uint8_t command[] = { CMD_READ, src_addr >> 16, src_addr >> 8, src_addr };
+    ice_ssram_input_command(cs_pin, command, sizeof(command), dest, size, false);
+}
+
+void ice_ssram_read_async(int cs_pin, void* dest, uint32_t src_addr, uint32_t size) {
+    uint8_t command[] = { CMD_READ, src_addr >> 16, src_addr >> 8, src_addr };
+    ice_ssram_input_command(cs_pin, command, sizeof(command), dest, size, true);
+}
+
+void ice_ssram_enable_power(int cs_pin, bool enabled) {
+    const uint8_t command[] = { enabled ? CMD_RELEASE_POWER_DOWN : CMD_POWER_DOWN };
+    ice_ssram_input_command(cs_pin, command, sizeof(command), NULL, 0, false);
+
+    if (enabled) {
+        sleep_us(5); // Command takes 3us per datasheet to take effect
+    }
+}
```

Comment:
Thank you for all these commands!
The fact that the `cs_pin` is a parameter would allow that library to be rather generic, move from "sram" to "smem" maybe?

Reply:
Yes I'll do the big rename next. "smem" sounds good.